service/raft/group0_voter_registry.cc

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include "group0_voter_registry.hh"
+
+#include <queue>
+
+#include <seastar/util/log.hh>
+
+#include "service/topology_state_machine.hh"
+
+namespace service {
+
+seastar::logger rvlogger("group0_voter_registry");
+
+future<> group0_voter_registry::insert_nodes(const std::unordered_set<raft::server_id>& nodes, abort_source& as) {
+    return update_voters(nodes, {}, as);
+}
+
+future<> group0_voter_registry::remove_nodes(const std::unordered_set<raft::server_id>& nodes, abort_source& as) {
+    return update_voters({}, nodes, as);
+}
+
+future<> group0_voter_registry::update_voters(
+        const std::unordered_set<raft::server_id>& nodes_added, const std::unordered_set<raft::server_id>& nodes_removed, abort_source& as) {
+    std::unordered_set<raft::server_id> voters_add;
+    std::unordered_set<raft::server_id> voters_del;
+    voters_add.reserve(nodes_added.size());
+    voters_del.reserve(nodes_removed.size());
+
+    // Consider all the new nodes non-voters first
+    for (const auto& node : nodes_added) {
+        const auto& server_info = _server_info_accessor.find(node);
+        auto& voters = _dc_voters[server_info.datacenter];
+        const auto itr = std::find_if(voters.begin(), voters.end(), [node](const auto& voter) {
+            return voter.second == node;
+        });
+        if (itr != voters.end()) {
+            rvlogger.warn("Node {} to be added is already present in the raft voter registry", node);
+            continue;
+        }
+        _dc_voters[server_info.datacenter].emplace(false, node);
+        ++_nodes_count;
+    }
+
+    // Handle the removed nodes
+    for (const auto& node : nodes_removed) {
+        // Make sure the node is always marked a non-voter
+        voters_del.emplace(node);
+
+        const auto& server_info = _server_info_accessor.find(node);
+        auto& voters = _dc_voters[server_info.datacenter];
+        const auto itr = std::find_if(voters.begin(), voters.end(), [node](const auto& voter) {
+            return voter.second == node;
+        });
+        if (itr == voters.end()) {
+            rvlogger.warn("Node {} to be removed not found in the raft voter registry", node);
+            continue;
+        }
+        voters.erase(itr);
+        --_nodes_count;
+
+        // Remove the DC if there are no nodes left
+        if (voters.empty()) {
+            _dc_voters.erase(server_info.datacenter);
+        }
+    }
+
+    // Distribute the available voter slots across the datacenters
+    auto slots_left_per_dc = distribute_slots();
```

Comment:
nit: I would call this method `distribute_slots_across_dcs` and remove this comment.

Reply:
Never mind, I see there is a commit with logic for racks